de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,543 @@
+                "referencePoint "
+                    + referencePoint
+                    + " is already completely shared (cannot downgrade a completely shared referencePoint)");
+
+        if (!isPartially
+            && partiallySharedReferencePoints.contains(referencePoint)) {
+            partiallySharedReferencePoints.remove(referencePoint);
+            upgrade = true;
+        }
+
+        if (isPartially)
+            partiallySharedReferencePoints.add(referencePoint);
+        else
+            completelySharedReferencePoints.add(referencePoint);
+
+        assert Collections.disjoint(completelySharedReferencePoints,
+            partiallySharedReferencePoints);
+
+        if (upgrade) {
+            // release resources
+            partiallySharedResourceMapping.put(referencePoint, null);
+
+            LOG.debug("upgraded partially shared referencePoint "
+                + referencePoint + " with ID " + id
+                + " to a completely shared referencePoint");
+            return;
+        }
+
+        idToReferencePointMapping.put(id, referencePoint);
+        referencePointToIDMapping.put(referencePoint, id);
+
+        if (isPartially)
+            partiallySharedResourceMapping.put(referencePoint,
+                new HashSet<IResource>());
+        else
+            partiallySharedResourceMapping.put(referencePoint, null);
+
+        LOG.debug("added referencePoint " + referencePoint + " with ID " + id
+            + " [completely shared:" + !isPartially + "]");
+    }
+
+    /**
+     * Removes a reference point from the set of currently shared
+     * referencePoint. Does nothing if the referencePoint is not shared.
+     * 
+     * @param id
+     *            the ID of the referencePoint to remove
+     */
+    public synchronized void removeReferencePoint(String id) {
+        IReferencePoint referencePoint = idToReferencePointMapping.get(id);
+
+        if (referencePoint == null) {
+            LOG.warn("could not remove referencePoint, no referencePoint is registerid with ID: "
+                + id);
+            return;
+        }
+
+        if (partiallySharedReferencePoints.contains(referencePoint))
+            partiallySharedReferencePoints.remove(referencePoint);
+        else
+            completelySharedReferencePoints.remove(referencePoint);
+
+        idToReferencePointMapping.remove(id);
+        referencePointToIDMapping.remove(referencePoint);
+        partiallySharedResourceMapping.remove(referencePoint);
+
+        LOG.debug("removed referencePoint " + referencePoint + " with ID " + id);
+    }
+
+    /**
+     * Adds the given resources to a <b>partially</b> shared reference point.
+     * 
+     * @param referencePoint
+     *            a referencePoint that was added as a partially shared
+     *            referencePoint
+     * @param resources
+     *            the resources to add
+     */
+    /*
+     * TODO needs proper sync. in the SarosSession class
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=210&issues=AWVWjH4plRPS_ea-bThV&open=AWVWjH4plRPS_ea-bThV)

[ISSUE_KEY]: AWVWjH4plRPS_ea-bThV